javaee/impl/src/main/java/org/jboss/forge/addon/javaee/jpa/ui/setup/JPASetupWizardImpl.java

```diff
@@ -82,32 +84,58 @@ public Metadata getMetadata(UIContext context)
    @Override
    public void initializeUI(UIBuilder builder) throws Exception
    {
-      initContainers(builder.getUIContext());
+      UIContext uiContext = builder.getUIContext();
+      if (initJpaVersion(uiContext))
+      {
+         builder.add(jpaVersion);
+      }
+      initContainers(uiContext);
       initProviders();
       initConfigureMetadata();
-      builder.add(jpaVersion).add(container).add(provider).add(configureMetadata);
+      builder.add(container).add(provider).add(configureMetadata);
    }
 
-   private void initContainers(UIContext context)
+   @SuppressWarnings({ "rawtypes", "unchecked" })
+   private boolean initJpaVersion(UIContext context)
    {
-      final boolean isGUI = context.getProvider().isGUI();
-      container.setItemLabelConverter(new Converter<PersistenceContainer, String>()
+      final Project project = getSelectedProject(context);
+      if (project.hasFacet(JPAFacet.class))
       {
-         @Override
-         public String convert(PersistenceContainer source)
-         {
-            return source != null ? source.getName(isGUI) : null;
-         }
-      });
-      // Ordering items:
-      TreeSet<PersistenceContainer> treeSet = new TreeSet<>(new Comparator<PersistenceContainer>()
+         // If JPA is already installed, do not ask for JPA Version
+         jpaVersion.setEnabled(false);
```

Comment:
This could perhaps be simplified via a @StackEnabled(JPAFacet.class) annotation on the UIInput itself.